### PR TITLE
SIngle-pass `CopyValueClassUnchecked`

### PR DESCRIPTION
**Pull request by VSadov**
Currently for structs with GC references `CopyValueClassUnchecked` does two passes - 1) to copy all data 2) to copy reference fields individually, doing fences and cards for each.

- It makes reference assignments not atomic. I.E for two reference fields `a` and `b`, assignments will be like - `a` is copied, `b` is copied, then `a` is copied again then `b` is copied again. 
Although it is the same values so it is hard to see someone broken by this. We do not guarantee the order of elements/fields updates while copying anyways.
- It is redundant in terms of number of copies and barriers executed.

We have a helper that does the same in a more economic way:  1-pass copy + bulk update cards - `memmoveGCRefs`. We already use that to copy arrays and for cloning objects. We should use that for struct copying too, for the same benefits.

### CODE REVIEW

File: src/coreclr/src/vm/object.cpp [hunk 1]
```diff
@@ -335,62 +335,43 @@ void STDCALL CopyValueClassUnchecked(void* dest, void* src, MethodTable *pMT)
...
-        CGCDescSeries* cur = map->GetHighestSeries();
-        CGCDescSeries* last = map->GetLowestSeries();
-        DWORD size = pMT->GetBaseSize();
-        _ASSERTE(cur >= last);
-        do
+        // <TODO> @todo Only call MemoryBarrier() if needed.
+        // Reflection is a known use case where this is required.
+        // Unboxing is a use case where this should not be required.
+        // </TODO>
+        MemoryBarrier();
```
Comment (jkotas) [inline src/coreclr/src/vm/object.cpp:348]: Should this be on the ContainsPointers path as well?

**Comment by jkotas** [comment]
Do you have any numbers for what this is improving?

File: src/coreclr/src/vm/object.cpp (hunk 1 above)
Comment (VSadov) [reply src/coreclr/src/vm/object.cpp:348]: ContainsPointers has a similar barrier before it does actual copying. 

I do not know why this is here though. A full fence is typically paired to a similar one which may run on a different thread, or something to a similar effect. The comment does not where that is, just says it is needed, so I kept it. 

The barrier in ContainsPointers path is a bit strange too. - If we would be copying element-wise in a for loop, we would not be doing full fence, so I am not sure.

**Comment by VSadov** [comment]
I noticed that all other bulk-copying scenarios were routed to `memmoveGCRefs` so I figured either this was missed and it would be an improvement or there is a chance this is special in some way and should be as-is.

I will try to measure. If it is worse we may need to rethink the use in object cloning as well.

**Comment by VSadov** [comment]
It is pretty hard to make jit to call this directly, perhaps not possible. Even unboxing will try to fetch a pointer and then do element-wise copy.
It looks like jit always emits element-wise copy, even for large structs. On samples that I tried jitted element-wise copy seems to perform better. There could be a size where using the helper might be a win, but it seems like it must be pretty big and will vary with environment so JIT strategy of inlining the copy seems reasonable.

Scenarios which are not jitted, like GetValue via reflection do use the helper and see improvements.

**Comment by VSadov** [comment]
I am using the following microbenchmark:
```cs
namespace StructCopy
{
    struct Small
    {
        public Small(string s)
        {
            s1 = s2 = s3 = s4 = s5 = s6 = s7 = s8 = s9 = s;
            x1 = x2 = x3 = x4 = x5 = x6 = x7 = x8 = x9 = 42;
        }

        string s1;
        int x1;
        string s2;
        int x2;
        string s3;
        int x3;
        string s4;
        int x4;
        string s5;
        int x5;
        string s6;
        int x6;
        string s7;
        int x7;
        string s8;
        int x8;
        string s9;
        int x9;
    }

    struct Large
    {
        public Large(string s)
        {
            s01 = new Small(s);
            s02 = new Small(s);
            s03 = new Small(s);
            s04 = new Small(s);
            s05 = new Small(s);
            s06 = new Small(s);
            s07 = new Small(s);
            s08 = new Small(s);
            s09 = new Small(s);
            s10 = new Small(s);
            s11 = new Small(s);
            s12 = new Small(s);
            s13 = new Small(s);
            s14 = new Small(s);
            s15 = new Small(s);
            s16 = new Small(s);
        }

        public Small s01;
        public Small s02;
        public Small s03;
        public Small s04;
        public Small s05;
        public Small s06;
        public Small s07;
        public Small s08;
        public Small s09;
        public Small s10;
        public Small s11;
        public Small s12;
        public Small s13;
        public Small s14;
        public Small s15;
        public Small s16;
    }

    class Program
    {
        public Small si;
        public static Small ss;

        public Large li;
        public static Large ls;

        static void Main(string[] args)
        {
            const int iterations = 1 << 25;
            Program p = new Program();

            p.si = new Small("small");
            p.li = new Large("large");

            var smallField = typeof(Program).GetField("si");
            var largeField = typeof(Program).GetField("li");

            Console.WriteLine("Small GetValue");

            Stopwatch sw = Stopwatch.StartNew();           
            for (long i = 1; i < iterations; i++)
            {
                if (i % (1<<22) == 0)
                {
                    System.Console.WriteLine(sw.ElapsedMilliseconds);
                    sw.Restart();
                }

                ss = (Small)smallField.GetValue(p);
            }

            Console.WriteLine("Small Unbox");
            object o = p.si;

            sw = Stopwatch.StartNew();
            for (long i = 1; i < iterations; i++)
            {
                if (i % (1 << 22) == 0)
                {
                    System.Console.WriteLine(sw.ElapsedMilliseconds);
                    sw.Restart();
                }

                ss = (Small)o;
            }

            Console.WriteLine("Large GetValue");

            sw = Stopwatch.StartNew();
            for (long i = 1; i < iterations; i++)
            {
                if (i % (1 << 22) == 0)
                {
                    System.Console.WriteLine(sw.ElapsedMilliseconds);
                    sw.Restart();
                }

                ls = (Large)largeField.GetValue(p);
            }

            Console.WriteLine("Large Unbox");
            o = p.li;

            sw = Stopwatch.StartNew();
            for (long i = 1; i < iterations; i++)
            {
                if (i % (1 << 22) == 0)
                {
                    System.Console.WriteLine(sw.ElapsedMilliseconds);
                    sw.Restart();
                }

                ls = (Large)o;
            }

        }

    }
}
```

**Comment by VSadov** [comment]
Results on x64 (Windows):
```
=== Baseline:
Small GetValue
428
420
419
417
421
418
417
Small Unbox
112
108
108
109
110
110
111
Large GetValue
3123
3065
2852
2860
2855
2853
2877
Large Unbox
1557
1557
1553
1562
1568
1563
1557
```

```
=== With this change:
Small GetValue
377
363
360
358
361
366
365
Small Unbox
104
108
104
100
117
107
103
Large GetValue
2171
2168
2180
2196
2192
2209
2196
Large Unbox
1487
1490
1492
1485
1491
1490
1491
```

I do not think Unbox path calls the helper. It is here just for reference.

**Comment by VSadov** [comment]
Results on ARM64 (Linux)

```
=== Baseline:
Small GetValue
2661
2639
2639
2550
2262
2262
2263
Small Unbox
419
419
419
419
419
419
419
Large GetValue
19108
19098
19098
19098
19099
19098
19098
Large Unbox
6796
6796
6796
6796
6796
6796
6796
```

```
=== With this change
Small GetValue
1720
1701
1701
1702
1702
1694
1675
Small Unbox
452
452
452
452
452
452
452
Large GetValue
8855
8859
8858
8858
8869
8862
8866
Large Unbox
6833
6833
6833
6833
6833
6833
6833
```

**Comment by VSadov** [comment]
Regarding the full fence. I think it was there to compensate for not doing write fences when copying the data.
Element-wise copy gets that because reference fields are copied via GC write barriers. `memcpy` does not get that, and since we did `memcpy` unconditionally, the fence was there unconditionally.

Based on this I think:
- we do not need any fences on  !ContainsPointers path.   
(indeed, element-wise copying would not have any)
- the full fence on ContainsPointers path (inside the `memmoveGCRefs`) is only needed on ARM. 
(again, on x64 element-wise copy would not have any fences)

**Review by jkotas — APPROVED** [review]

**Comment by VSadov** [comment]
Thanks!!

**Pull request merged**